### PR TITLE
Devb

### PR DESCRIPTION
**Pull request by benroywillis**
Working Set Tool, TikSwap, Comments

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -64,9 +64,10 @@ add_subdirectory("tik")
 add_subdirectory("deat")
 add_subdirectory("kar")
 add_subdirectory("Utilities")
+add_subdirectory("WorkingSet")
```
Comment (ruhrie) [inline CMakeLists.txt]: This honestly needs to move into the util subdirectory. It will be time the dag extractor to move as well as soon as this tool or @chang55245 have a tool that fully supports kernel instances.

Comment (ruhrie) [reply CMakeLists.txt]: Also, it needs a name to differentiate it. It isn't the only working set tool in development as far as I am aware.

File: Utilities/CMakeLists.txt [hunk 2]
```diff
@@ -11,8 +11,9 @@ add_executable(kernelHasher KernelHasher.cpp)
 
 set_target_properties(kernelHasher PROPERTIES RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/bin")
 target_compile_definitions(kernelHasher PUBLIC ${LLVM_DEFINITIONS})
-target_link_libraries(kernelHasher ${llvm_libs} nlohmann_json nlohmann_json::nlohmann_json)
+target_link_libraries(kernelHasher ${llvm_libs} nlohmann_json nlohmann_json::nlohmann_json libtik AtlasUtil)
```
Comment (ruhrie) [inline Utilities/CMakeLists.txt]: Kernel hasher should no longer require libtik.

File: Utilities/CMakeLists.txt [hunk 3]
```diff
@@ -11,8 +11,9 @@ add_executable(kernelHasher KernelHasher.cpp)
...
 target_include_directories(kernelHasher SYSTEM PRIVATE ${LLVM_INCLUDE_DIRS})
+target_include_directories(kernelHasher PRIVATE "${PROJECT_SOURCE_DIR}/tik/lib/include/")
```
Comment (ruhrie) [inline Utilities/CMakeLists.txt]: This include should be added through an interface property.

File: Utilities/KernelHasher.cpp [hunk 4]
```diff
@@ -38,34 +43,39 @@ int main(int argc, char **argv)
...
+    try
+    {
+        inputJson.open(KernelFilename);
+        inputJson >> j;
+        inputJson.close();
+    }
+    catch (exception &e)
+    {
+        std::cerr << "Couldn't open input json file: " << KernelFilename << "\n";
+        std::cerr << e.what() << '\n';
```
Comment (ruhrie) [inline Utilities/KernelHasher.cpp]: change to spdlog

File: WorkingSet/WorkingSetAnalysis.cpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            spdlog::set_level(spdlog::level::debug);
+        }
+        case 6:
+        {
+            spdlog::set_level(spdlog::level::trace);
+            break;
+        }
+        default:
+        {
+            spdlog::warn("Invalid logging level: " + to_string(LogLevel));
```
Comment (ruhrie) [inline WorkingSet/WorkingSetAnalysis.cpp]: All cases, including default should have a break. If mine didn't, that was my mistake. Regardless, it should be fixed.

File: WorkingSet/WorkingSetAnalysis.cpp [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    for (const auto &ind : WorkingSet::ProdConRelationships)
+    {
+        outputJson["Producer-Consumer"]["Kernel IDs"][to_string(ind.kernels.first) + "," + to_string(ind.kernels.second)]["Input,Output"] = ind.InputOutput.size();
+        outputJson["Producer-Consumer"]["Kernel IDs"][to_string(ind.kernels.first) + "," + to_string(ind.kernels.second)]["Internal,Internal"] = ind.InternalInternal.size();
+        outputJson["Producer-Consumer"]["Kernel IDs"][to_string(ind.kernels.first) + "," + to_string(ind.kernels.second)]["Output,Input"] = ind.OutputInput.size();
+    }
+    ofstream out(OutputFile);
+    out << setw(4) << outputJson << endl;
+    out.close();
+    return 0;
```
Comment (ruhrie) [inline WorkingSet/WorkingSetAnalysis.cpp]: EXIT_SUCCESS please

File: tik/lib/TikKernel.cpp [hunk 7]
```diff
@@ -1,15 +1,17 @@
 #include "tik/TikKernel.h"
 #include "AtlasUtil/Exceptions.h"
+#include <iostream>
```
Comment (ruhrie) [inline tik/lib/TikKernel.cpp:3]: Why is iostream here?

File: tik/lib/Util.cpp [hunk 8]
```diff
@@ -13,24 +14,91 @@ using namespace llvm;
...
+                {
+                    RecurseThroughOperands(subVal);
+                }
+            }
+        }
+        else if (auto go = dyn_cast<GlobalObject>(val))
+        {
+            if (GetValueID(go) < 0)
+            {
+                PrintVal(go);
```
Comment (ruhrie) [inline tik/lib/Util.cpp]: Should only be used for debugging. If this is ready to merge it should be removed.

File: tik/lib/Util.cpp [hunk 9]
```diff
@@ -449,4 +486,68 @@ namespace TraceAtlas::tik
...
+                        if (s.find(parentBlock) == s.end() && parentBlock->getParent() == entrance->getParent())
+                        {
+                            result.insert(parentBlock);
+                        }
+                    }
+                }
+            }
+        }
+        return result;
+    }
```
Comment (ruhrie) [inline tik/lib/Util.cpp:464]: Did you rewrite the exit code? github marks the entire area as a delta.

If so, this code has been migrated to AtlasUtil.

File: tik/lib/include/tik/TikKernel.h [hunk 10]
```diff
@@ -2,10 +2,15 @@
...
+    ///
+    /// The constructor of this class will either construct a TikKernel if the given argument is actually a kernel function, or will return an invalid TikKernel
+    /// A function is only a valid TikKernel if it has Boundaries, Entrances, and Exits metadata attached to it.
+    /// If the function is valid, the constructor will allocate a new function with the context of the input arg.
     class TikKernel : public Kernel
     {
     public:
         TikKernel() = default;
-        TikKernel(llvm::Function *kernFunc);
+        TikKernel(llvm::Function *);
```
Comment (ruhrie) [inline tik/lib/include/tik/TikKernel.h:14]: Why did you remove this argument name?

Please avoid cosmetic changes like this if there is no reason. It makes the diffs harder to read.

File: tik/lib/include/tik/Util.h [hunk 11]
```diff
@@ -22,9 +29,10 @@ namespace TraceAtlas::tik
     std::set<llvm::BasicBlock *> GetReachable(llvm::BasicBlock *base, std::set<int64_t> validBlocks);
     bool IsSelfReachable(llvm::BasicBlock *base, const std::set<int64_t> &validBlocks);
     bool IsReachable(llvm::BasicBlock *base, llvm::BasicBlock *target, const std::set<int64_t> &validBlocks);
-    std::set<llvm::BasicBlock *> GetEntrances(std::set<llvm::BasicBlock *> &blocks);
-    std::set<llvm::BasicBlock *> GetExits(std::set<llvm::BasicBlock *> blocks);
-    std::set<llvm::BasicBlock *> GetExits(llvm::Function *F);
+    std::set<llvm::BasicBlock *> GetEntrances(const std::set<llvm::BasicBlock *> &);
+    std::set<llvm::BasicBlock *> GetExits(std::set<llvm::BasicBlock *> &, llvm::BasicBlock *);
+    std::set<llvm::BasicBlock *> GetExits(std::set<llvm::BasicBlock *>);
+    std::set<llvm::BasicBlock *> GetExits(llvm::Function *);
```
Comment (ruhrie) [inline tik/lib/include/tik/Util.h:35]: Same question, why were all the argument names removed?

Comment (ruhrie) [reply tik/lib/include/tik/Util.h:35]: The names and the types are the only bits passed to doxygen. With them removed it makes these functions very anonymous.

File: tik/tik.cpp [hunk 12]
```diff
@@ -332,7 +337,7 @@ int main(int argc, char *argv[])
     }
     if (error)
     {
-        return EXIT_FAILURE;
```
Comment (ruhrie) [inline tik/tik.cpp:339]: Don't we still want this to report as a failure?

File: tik/lib/include/tik/Kernel.h [hunk 13]
```diff
@@ -22,21 +22,38 @@ namespace TraceAtlas::tik
...
+            }
+            bool operator()(const std::shared_ptr<KernelInterface> &lhs, int64_t rhs) const
+            {
+                return lhs->Block < rhs;
+            }
+            bool operator()(int64_t lhs, const std::shared_ptr<KernelInterface> &rhs) const
+            {
+                return lhs < rhs->Block;
+            }
+        };
```
Comment (ruhrie) [inline tik/lib/include/tik/Kernel.h:40]: I'm confused. Why is this necessary? Can this not just be built into the KernelInterface?

I don't understand what is happening here.

File: tik/tikSwap/tikSwap.cpp [hunk 14]
```diff
@@ -32,53 +34,40 @@ cl::opt<bool> ASCIIFormat("S", cl::desc("output json as human-readable ASCII tex
...
-                if (md->getNumOperands() > 0)
-                {
-                    if (auto con = mdconst::dyn_extract<ConstantInt>(md->getOperand(0)))
-                    {
-                        int64_t blockID = con->getSExtValue();
-                        baseBlockMap[blockID] = block;
-                    }
-                }
-            }
+            F.setLinkage(GlobalValue::ExternalLinkage);
```
Comment (ruhrie) [inline tik/tikSwap/tikSwap.cpp:72]: Is this correct? I suspect this may break with code where we only link against some symbols, but not all. I'm thinking about where we still have to link to some so libraries.

Trying to start the conversation. Not sure what the answer is.

File: tik/tikSwap/tikSwap.cpp [hunk 15]
```diff
@@ -88,196 +77,317 @@ int main(int argc, char *argv[])
...
-        auto kern = new TikKernel(func);
-        if (kern->Valid)
-        {
-            kernels.push_back(kern);
-        }
-        else
+        if (func.hasMetadata("Boundaries"))
         {
-            delete kern;
+            auto kern = new TikKernel(tikBitcode->getFunction(func.getName()));
```
Comment (ruhrie) [inline tik/tikSwap/tikSwap.cpp]: Please avoid new and delete like the plague. Use shared_ptr or unique_ptr depending on your context.

shared_ptr is probably the right answer here.

File: tik/tikSwap/tikSwap.cpp [hunk 16]
```diff
@@ -88,196 +77,317 @@ int main(int argc, char *argv[])
...
+                                    }
+                                    // do not do memory operations in the entrance block, unless the user is a phi
+                                    if (inst->getParent() == IDToBlock[e->Block] && dyn_cast<PHINode>(inst) == nullptr)
+                                    {
+                                        continue;
+                                    }
+                                    toReplace.emplace_back(pair(IDToValue[key.second], inst));
+                                }
+                            }
+                            KInst->replaceUsesOfWith(IDToValue[key.second], alloc);
```
Comment (ruhrie) [inline tik/tikSwap/tikSwap.cpp:285]: Just a comment, I've found uses of this function to be problematic. If it is working great, but it has caused numerous bugs for me in the past. Just keep an eye on it.

File: tik/tikSwap/tikSwap.cpp [hunk 17]
```diff
@@ -88,196 +77,317 @@ int main(int argc, char *argv[])
...
+            }
+        }
+    }
 
-    //verify the module
+    // verify the module
     std::string str;
     llvm::raw_string_ostream rso(str);
-    bool broken = verifyModule(*base, &rso);
+    bool broken = verifyModule(*sourceBitcode, &rso);
```
Comment (ruhrie) [inline tik/tikSwap/tikSwap.cpp]: Is this code that would be best moved to AtlasUtil?

We seem to be checking for broken modules with increasing frequency.

File: tik/tikSwap/tikSwap.cpp [hunk 18]
```diff
@@ -307,14 +417,14 @@ int main(int argc, char *argv[])
             f.open(OutputFile, std::ios::out);
             std::ostream rawStream(&f);
             raw_os_ostream raw_stream(rawStream);
-            WriteBitcodeToFile(*base, raw_stream);
+            WriteBitcodeToFile(*sourceBitcode, raw_stream);
         }
-        spdlog::info("Successfully wrote tik to file");
+        spdlog::info("Successfully wrote tikSwap to file");
```
Comment (ruhrie) [inline tik/tikSwap/tikSwap.cpp]: Same on the print. I suspect this will see increased use in more tools. Thoughts?

File: tik/lib/CartographerKernel.cpp [hunk 19]
```diff
@@ -192,95 +220,57 @@ namespace TraceAtlas::tik
...
-                            throw AtlasException("Null function call (indirect call)");
+                            bi.moveAfter(st);
                         }
-
-                        auto *funcDec = cast<Function>(TikModule->getOrInsertFunction(callBase->getCalledFunction()->getName(), callBase->getCalledFunction()->getFunctionType()).getCallee());
-                        funcDec->setAttributes(callBase->getCalledFunction()->getAttributes());
-                        callBase->setCalledFunction(funcDec);
                     }
                 }
             }
```
Comment (ruhrie) [inline tik/lib/CartographerKernel.cpp:258]: I'm not sure what this code is doing. It used to be a catch for indirect function calls, but that appears to have been removed and replaced with an inlining check?

File: tik/lib/CartographerKernel.cpp [hunk 20]
```diff
@@ -291,500 +281,926 @@ namespace TraceAtlas::tik
...
-                            if (find(KernelImports.begin(), KernelImports.end(), GetValueID(ar)) == KernelImports.end())
+                            if (scopedFuncs.find(operand->getParent()->getParent()) == scopedFuncs.end())
                             {
-                                KernelImports.push_back(GetValueID(ar));
+                                auto sExtVal = GetValueID(operand);
+                                if (kernelIE.find(sExtVal) == kernelIE.end())
+                                {
+                                    KernelImports.push_back(sExtVal);
+                                    kernelIE.insert(sExtVal);
+                                }
```
Comment (ruhrie) [inline tik/lib/CartographerKernel.cpp:422]: You no longer check for arguments. Is there a reason you can ignore them?

File: tik/lib/CartographerKernel.cpp [hunk 21]
```diff
@@ -291,500 +281,926 @@ namespace TraceAtlas::tik
...
+                                        throw AtlasException("Import needs to be an export!");
+                                    }
+                                }
+                            }
+                        }
                     }
                 }
             }
         }
-        if (Entrances.empty())
```
Comment (ruhrie) [inline tik/lib/CartographerKernel.cpp:392]: This check is still very important until kernels are parsed correctly 100% of the time.

File: tik/lib/CartographerKernel.cpp [hunk 22]
```diff
@@ -291,500 +281,926 @@ namespace TraceAtlas::tik
...
-                        break;
-                    }
-                }
-                if (rescheduled > 1)
-                {
-                    spdlog::warn("Rescheduled more than one phi predecessor"); //basically this is a band aid. Needs some more help
                 }
             }
         }
     }
```
Comment (ruhrie) [inline tik/lib/CartographerKernel.cpp:591]: I really want some pseudo-code of what this is doing. I'm really struggling to read it due to the diff and do not quite understand what the problem is or what you are doing.

File: tik/lib/CartographerKernel.cpp [hunk 23]
```diff
@@ -291,500 +281,926 @@ namespace TraceAtlas::tik
...
+                        break;
+                    }
+                }
+                if (inNested)
                 {
-                    auto calledFunc = callInst->getCalledFunction();
-                    auto subK = KfMap[calledFunc];
-                    if (subK != nullptr)
+                    //we need to make a unique block for each entrance (there is currently only one)
+                    for (uint64_t i = 0; i < nestedKernel->Entrances.size(); i++)
```
Comment (ruhrie) [inline tik/lib/CartographerKernel.cpp:743]: The ideal answer is to create a select statement that the phi will select if either entrance is used.

As long as this work we can leave it though. I think the optimizer should fix it ultimately anyway.

File: tik/lib/CartographerKernel.cpp [hunk 24]
```diff
@@ -291,500 +281,926 @@ namespace TraceAtlas::tik
...
+
+        for (auto fi = KernelFunction->begin(); fi != KernelFunction->end(); fi++)
+        {
+            if (auto b = dyn_cast<BasicBlock>(fi))
+            {
+                //do nothing
+            }
+            else
+            {
+                toRemove.push_back(fi);
```
Comment (ruhrie) [inline tik/lib/CartographerKernel.cpp:1182]: what are these non-blocks?

File: tik/lib/CartographerKernel.cpp [hunk 25]
```diff
@@ -952,166 +1310,346 @@ namespace TraceAtlas::tik
...
-            for (auto bi = fi->begin(); bi != fi->end(); bi++)
+            for (auto bi = fi.begin(); bi != fi.end(); bi++)
             {
                 if (auto ii = dyn_cast<InvokeInst>(bi))
                 {
                     auto a = ii->getLandingPadInst();
                     if (isa<BranchInst>(a))
                     {
-                        auto unwind = ii->getUnwindDest();
+                        throw AtlasException("Exception handling is not supported");
```
Comment (ruhrie) [inline tik/lib/CartographerKernel.cpp:1643]: This almost certainly removes support for all invoke instructions.

**Review by ruhrie — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (benroywillis) [reply CMakeLists.txt]: Fixed

File: Utilities/CMakeLists.txt (hunk 2 above)
Comment (benroywillis) [reply Utilities/CMakeLists.txt]: It does now. I use a function here to map all basic blocks to their respective ID (InitializeIDMaps()), and that symbol is from tik/Util.cpp. If you want this, I want to move that function to AtlasUtil/Annotate.h and then I can remove the libtik dependency

File: Utilities/CMakeLists.txt (hunk 3 above)
Comment (benroywillis) [reply Utilities/CMakeLists.txt]: Going to hold off on this until we agree on the above comment

File: Utilities/KernelHasher.cpp (hunk 4 above)
Comment (benroywillis) [reply Utilities/KernelHasher.cpp]: fixed

File: WorkingSet/WorkingSetAnalysis.cpp (hunk 5 above)
Comment (benroywillis) [reply WorkingSet/WorkingSetAnalysis.cpp]: fixed

File: WorkingSet/WorkingSetAnalysis.cpp (hunk 6 above)
Comment (benroywillis) [reply WorkingSet/WorkingSetAnalysis.cpp]: fixed

File: tik/lib/CartographerKernel.cpp (hunk 19 above)
Comment (benroywillis) [reply tik/lib/CartographerKernel.cpp:258]: Ah, so this is misleading. These two things are actually not replacements of each other, even though the diffs appear to be so.

The check for indirect function calls has been moved to FindScopedStructures() at the top of this file. Therefore function pointers don't need to be checked here.

Now this check is on instructions created by me that have accidentally found themselves as the block terminator (my case test is for Halide/fft program where a store instruction to an export is for some reason a block terminator).

File: tik/lib/CartographerKernel.cpp (hunk 21 above)
Comment (benroywillis) [reply tik/lib/CartographerKernel.cpp:392]: This check has been moved to the top of ConstructFunctionSignature()

File: tik/lib/CartographerKernel.cpp (hunk 20 above)
Comment (benroywillis) [reply tik/lib/CartographerKernel.cpp:422]: That check is now further up, right after the error handling if() block

File: tik/lib/CartographerKernel.cpp (hunk 22 above)
Comment (benroywillis) [reply tik/lib/CartographerKernel.cpp:591]: The description of GetBoundaryValues() summarizes the core of this routine. It is the naming, gathering of all function arguments and all the metadata (arg names) and error handling (bad args) that go along with that, and a mapping of original values to imiport args in the VMap.
Name -> function args gathering -> function object construction, arg naming and mapping, error checks -> VMap of the imports

File: tik/lib/CartographerKernel.cpp (hunk 23 above)
Comment (benroywillis) [reply tik/lib/CartographerKernel.cpp:743]: You're saying we should create 1 block that enters the child from all entrances of a child, instead of a unique block for all entrances of a child

File: tik/lib/CartographerKernel.cpp (hunk 24 above)
Comment (benroywillis) [reply tik/lib/CartographerKernel.cpp:1182]: I dont believe I contributed this, I think it appears green because i moved this stuff around. I don't know if its actually needed.

File: tik/lib/CartographerKernel.cpp (hunk 25 above)
Comment (benroywillis) [reply tik/lib/CartographerKernel.cpp:1643]: it does and thats by design. When this code is exercised and results in the generation of invoke instructions, there is a module error, either because the personality function is from another module, or because the invoke instruction uses a null type, which is not allowed

File: tik/lib/TikKernel.cpp (hunk 7 above)
Comment (benroywillis) [reply tik/lib/TikKernel.cpp:3]: shouldn't be. gone now

File: tik/lib/Util.cpp (hunk 8 above)
Comment (benroywillis) [reply tik/lib/Util.cpp]: fixed

File: tik/lib/Util.cpp (hunk 9 above)
Comment (benroywillis) [reply tik/lib/Util.cpp:464]: I added a function that is used when I collect function args in GetBounaryValues(). It creates an estimation for all exit blocks before we have the kernel function. Its purpose is to provide a set of blocks whose child instructions may have users outside the kernel via a premature exit. Thus the instructions belonging to blocks in that set are added to the export list

File: tik/lib/include/tik/Kernel.h (hunk 13 above)
Comment (benroywillis) [reply tik/lib/include/tik/Kernel.h:40]: This allows me to search the sets Entrances, Exits that have KernelInterface objects with block IDs only, not KernelInterface objects

File: tik/lib/include/tik/TikKernel.h (hunk 10 above)
Comment (benroywillis) [reply tik/lib/include/tik/TikKernel.h:14]: I changed these things because every time I change a function signature I have to do it twice. When theres a change to an arg name, I dont have to do it twice. Though I'll add these back anyway

File: tik/lib/include/tik/Util.h (hunk 11 above)
Comment (benroywillis) [reply tik/lib/include/tik/Util.h:35]: Same reason as above, but ill change it back

File: tik/tik.cpp (hunk 12 above)
Comment (benroywillis) [reply tik/tik.cpp:339]: No. The new Dash-Automate flow pays attention to every return code. If an error is detected the pipeline is stopped. Thus, we need to use return codes to the OS to signify whether the tool succeeded in the defined-behavior sense. This error check is about a semantic error, and there needs to be signified in a different way.

File: tik/tikSwap/tikSwap.cpp (hunk 14 above)
Comment (benroywillis) [reply tik/tikSwap/tikSwap.cpp:72]: I found that the linker didn't reliably work when this was absent. Somethings objects used in the Tik module would be defined in the original source bitcode. But (at least) a lot of the symbols in the original are closed to the outside world, can't remember what the keyword is. So I change them to the default external linkage so that they're allowed to link in another module

File: tik/tikSwap/tikSwap.cpp (hunk 17 above)
Comment (benroywillis) [reply tik/tikSwap/tikSwap.cpp]: I'm alright with that. want me to do it?

File: tik/tikSwap/tikSwap.cpp (hunk 18 above)
Comment (benroywillis) [reply tik/tikSwap/tikSwap.cpp]: Yeah, that's a good idea

File: tik/tikSwap/tikSwap.cpp (hunk 16 above)
Comment (benroywillis) [reply tik/tikSwap/tikSwap.cpp:285]: haven't had a problem yet, I use it to target specific values that don't fit well into the VMap. Same case in CartographerKernel. When replacing the values of exports, we have a pointer mapping to its data. this can't just be mapped 1:1 in the VMap because the types don't match, and what I found was that the Remap() function changes the types of instructions when you remap with a different type, leading to problems.

File: tik/tikSwap/tikSwap.cpp (hunk 15 above)
Comment (benroywillis) [reply tik/tikSwap/tikSwap.cpp]: gave that a try initially. Can't remember the problem I had with it but I think it had to do with segfaulting when trying to access the shared pointer in the kernels container later. Didn't know how to fix it so reverted to this

File: Utilities/CMakeLists.txt (hunk 2 above)
Comment (ruhrie) [reply Utilities/CMakeLists.txt]: That's fine. It was honestly laziness earlier on my end where this started to bleed in. I believe I removed this type of mistake from most areas though, and we shouldn't add new ones.

File: tik/lib/CartographerKernel.cpp (hunk 23 above)
Comment (ruhrie) [reply tik/lib/CartographerKernel.cpp:743]: Ideally if there are multiple "Entrances" that effectively enter the same block we don't add additional blocks. This isn't for all entrances, just for entrances that share a target. As I said, this can be skipped, but may make the code harder to understand in both c++ and LLVM.

File: CMakeLists.txt (hunk 1 above)
Comment (ruhrie) [reply CMakeLists.txt]: There is no commit that fixed this. Did you push?

File: tik/lib/Util.cpp (hunk 9 above)
Comment (ruhrie) [reply tik/lib/Util.cpp:464]: Please check AtlasUtil and get back to me. An estimate should not be necessary. Is there a reason you can't use that code?

File: tik/lib/include/tik/TikKernel.h (hunk 10 above)
Comment (ruhrie) [reply tik/lib/include/tik/TikKernel.h:14]: The arg names don't have to match when compiling. They should match when doing a pull request though.

File: tik/lib/include/tik/Kernel.h (hunk 13 above)
Comment (ruhrie) [reply tik/lib/include/tik/Kernel.h:40]: I'm going to ask for a more distinct name or a comment explaining that then.

File: tik/tikSwap/tikSwap.cpp (hunk 15 above)
Comment (ruhrie) [reply tik/tikSwap/tikSwap.cpp]: The garbage collector disposes of the pointer as soon as it passes out of scope. If you want to maintain it put it in a vector in the lifetime of the object you are trying to store.

File: tik/tikSwap/tikSwap.cpp (hunk 17 above)
Comment (ruhrie) [reply tik/tikSwap/tikSwap.cpp]: Yes please.

File: tik/tikSwap/tikSwap.cpp (hunk 18 above)
Comment (ruhrie) [reply tik/tikSwap/tikSwap.cpp]: Let me know once its in.

File: tik/lib/CartographerKernel.cpp (hunk 25 above)
Comment (ruhrie) [reply tik/lib/CartographerKernel.cpp:1643]: The personality function is provided by libcpp. It shouldn't be defined in the other module. I would really like us to try and fix this.

**Comment by ruhrie** [comment]
I suspect you forgot to push.

Also there are three file conflicts that still need to be resolved.

File: Utilities/CMakeLists.txt (hunk 2 above)
Comment (benroywillis) [reply Utilities/CMakeLists.txt]: should be fixed now. Had to move around the declaration of IDToX Maps to fit within the new header structure

File: CMakeLists.txt (hunk 1 above)
Comment (benroywillis) [reply CMakeLists.txt]: should be up now

File: Utilities/CMakeLists.txt (hunk 3 above)
Comment (benroywillis) [reply Utilities/CMakeLists.txt]: should be done now

File: Utilities/KernelHasher.cpp (hunk 4 above)
Comment (benroywillis) [reply Utilities/KernelHasher.cpp]: should be fixed one

File: tik/lib/CartographerKernel.cpp (hunk 23 above)
Comment (benroywillis) [reply tik/lib/CartographerKernel.cpp:743]: considering this resolved

File: tik/lib/CartographerKernel.cpp (hunk 25 above)
Comment (benroywillis) [reply tik/lib/CartographerKernel.cpp:1643]: not what i've observed. a good example of this is opencv/contours2. we should work through it later.

File: tik/lib/Util.cpp (hunk 9 above)
Comment (benroywillis) [reply tik/lib/Util.cpp:464]: Before we acquire the exits of the kernel (because our current paradigm requires the kernel function to be constructed first), the function args have to be found. When a kernel can exit early, and the predecessors to that exit hold values that are used later in the kernel (and thus, after an early exit, there will be users of this value in the original bitcode), the value has to be exported or else it is unresolved at compile time

**Review by ruhrie — CHANGES_REQUESTED** [review]
If your working set tool requires an include or more than one source file please make it its own directory in util like nc has.

**Review by ruhrie — DISMISSED** [review]

**Review by ruhrie — APPROVED** [review]
